### PR TITLE
Add unconflict message id allocation

### PR DESCRIPTION
**Pull request by redboltz**

### CODE REVIEW

**Comment by redboltz** [comment]
This PR adds unique messageId provider option.
The current messageId allocating algorithm is starting with a random number between 1 to 65535, and then allocating incremental number. If the number reached to 65535 then back to 1.
The messageId conflict sometimes happened (not so often).

I implemented number-allocator.
See https://www.npmjs.com/package/number-allocator

And I implemented UniqueMessageIdProvider using number-allocator.
The original messageId allocationg algorithm is moved to DefaultNumberAllocator. So the this pull request doesn't change the default behavior. When user set the option `messageIdProvider: new UniqueMessageIdProvider()`, then UniqueMessageIdProvider is used instad of DefaultMessageIdProvider.

I have added tests.
node 10.x is failed but the master branch is also failed.
According to he README.md,
> now supports node v12 and v14.

So I think it is not a problem.

I don't add the document about the option yet.
I think that I respect the rule of MQTT.js but I'm not 100% sure that my naming rule for Classes and files are good for MQTT.js.

I'm ready to refine the pull request. Feedbacks are very welcome.

**Comment by redboltz** [comment]
I updated the PR.
Now, offline publish/subscribe/unsubscribe work well even if the messageId is run out.
The publish/subscribe/unsubscribe during messageId empry is delayed until messageId is reusable by puback/pubcomp/suback/unsuback receiving.

Type script support is refined.
Document is updated.

The PR is ready to review.

**Comment by redboltz** [comment]
NOTE:
This PR solves https://github.com/mcollina/mqtt-level-store/issues/19#issuecomment-560149375 .

**Comment by redboltz** [comment]
@YoDaMa , Thank you for restarting merging process!
After the many PRs merge, some conflicts happened on #1243 (this PR) . So I resolve them.

File: lib/client.js
Comment (YoDaMa) [inline lib/client.js:512]: I see here you're checking that it is null, would packet.messageId be null or would it be undefined?

**Comment by YoDaMa** [comment]
I'm going to hold off on merging this rn, so I can include it in the next release.  Once this current release is out and my comments are resolved we can merge.

File: lib/client.js
Comment (redboltz) [reply lib/client.js:512]: It is null.

Let me explain.
messageId is allocated using messageIdProvider.
See https://github.com/redboltz/MQTT.js/blob/add_unconflict_message_id_allocation/lib/client.js#L188
Note that the original implementation using `_nextId()`. I remained `_nextId()` but it uses `messageIdProvider.allocate()`
See https://github.com/redboltz/MQTT.js/blob/add_unconflict_message_id_allocation/lib/client.js#L1503-L1505

There are two types of messageIdProvider.
One is simply incremental. The initial value is random. It is the original behavior. The messageId could conflict. It never return null (or undefined).
See https://github.com/redboltz/MQTT.js/blob/add_unconflict_message_id_allocation/lib/default-message-id-provider.js#L25-L33
The other is allocating an unique number. 
See https://github.com/redboltz/MQTT.js/blob/add_unconflict_message_id_allocation/lib/unique-message-id-provider.js#L23-L26
If the all available ids are used, then return **null**, not undefined.
Implementation is here;
https://github.com/redboltz/number-allocator/blob/1.0.7/lib/number-allocator.js#L62-L84

See alloc() https://www.npmjs.com/package/number-allocator 
## alloc()
Allocate the first vacant number. The number become occupied status.

Time Complexity O(1)
- return: Number
  - The first vacant number. If all numbers are occupied, return null.

So the messageId in the client.js never be undefined.

I noticed that the comment is misleading.  I will update it.
https://github.com/redboltz/MQTT.js/blob/add_unconflict_message_id_allocation/lib/unique-message-id-provider.js#L21

**Comment by redboltz** [comment]
> I'm going to hold off on merging this rn, so I can include it in the next release. Once this current release is out and my comments are resolved we can merge.

Thank you for the comment. I answered your comment (null or undefined).
In addition, I updated the comment https://github.com/mqttjs/MQTT.js/pull/1243/commits/3907b67fc63c30564e47db50f242ccf0a5aa5cbe for easy understanding.

**Review by YoDaMa — APPROVED** [review]

**Pull request merged**
